Core/Modes/Matchmode.cs

```diff
@@ -1,38 +1,118 @@
+                Red = ImmutableList.Create(MatchTesting.TestVenonatForOverlay),
+            };
+            await Task.Delay(TimeSpan.FromSeconds(3));
+
+            IMatchCycle match = new CoinflipMatchCycle(_loggerFactory.CreateLogger<CoinflipMatchCycle>());
+            Task setupTask = match.SetUp(new MatchInfo(teams.Blue, teams.Red));
+            await _overlayConnection.Send(new MatchCreatedEvent());
+            await _overlayConnection.Send(new MatchBettingEvent());
+            await _overlayConnection.Send(new MatchModesChosenEvent()); // TODO
+            await _overlayConnection.Send(new MatchSettingUpEvent
+            {
+                MatchId = 1234,
+                Teams = teams,
+                BettingDuration = _matchmodeConfig.DefaultBettingDuration.TotalSeconds,
+                RevealDuration = 0,
+                Gimmick = "speed",
+                Switching = "never",
+                BattleStyle = "singles",
+                InputOptions = new InputOptions
+                {
+                    Moves = new MovesInputOptions
+                    {
+                        Policy = "always",
+                        Permitted = ImmutableList.Create("a", "b", "c", "d")
+                    },
+                    Switches = new SwitchesInputOptions
+                    {
+                        Policy = "never",
+                        Permitted = ImmutableList<string>.Empty,
+                        RandomChance = 0
+                    },
+                    Targets = new TargetsInputOptions
+                    {
+                        Policy = "disabled",
+                        Permitted = ImmutableList<string>.Empty,
+                        AllyHitChance = 0
+                    },
+                },
+                BetBonus = 35,
+                BetBonusType = "bet",
+            });
+
+            Duration bettingBeforeWarning = _matchmodeConfig.DefaultBettingDuration - _matchmodeConfig.WarningDuration;
+            await Task.Delay(bettingBeforeWarning.ToTimeSpan());
+            await _overlayConnection.Send(new MatchWarningEvent());
+
+            await Task.Delay(_matchmodeConfig.WarningDuration.ToTimeSpan());
+            await setupTask;
+            Task<MatchResult> performTask = match.Perform();
+            await _overlayConnection.Send(new MatchPerformingEvent { Teams = teams });
+
+            MatchResult result = await performTask;
+            object winnerForOverlay = result.Winner switch { Side.Blue => 0, Side.Red => 1, _ => "draw" };
+            await _overlayConnection.Send(new MatchOverEvent { MatchResult = winnerForOverlay });
+
+            await Task.Delay(_matchmodeConfig.ResultDuration.ToTimeSpan());
+            await _overlayConnection.Send(new ResultsFinishedEvent());
+        }
+
         public void Cancel()
         {
             // once the mainloop is not just busylooping, this needs to be replaced with something
```

Comment:
this needs to be done now

Reply:
so do you want to do this before I review or have a different PR for it?

Reply:
you can already look at it, but nothings blocking me from fixing that, I'm just procrastinating.
I'll likely just pass a cancellation token to Loop()